File: src/grove/discover.py
"""Discover git repositories in directories."""

from __future__ import annotations

import json
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass
from pathlib import Path

from grove.git import parse_remote_name, remote_url

# ---------------------------------------------------------------------------
# Remote URL cache — avoids re-spawning `git remote get-url` on every run
# ---------------------------------------------------------------------------

_CACHE_DIR = Path.home() / ".grove" / "cache"
_REMOTE_CACHE_FILE = _CACHE_DIR / "remotes.json"
_CACHE_TTL_SECONDS = 86400  # 24 h — remote URLs rarely change


def _load_remote_cache() -> dict[str, dict]:
    """Load the on-disk remote URL cache."""
    try:
        return json.loads(_REMOTE_CACHE_FILE.read_text())
    except (FileNotFoundError, json.JSONDecodeError, OSError):
        return {}


def _save_remote_cache(cache: dict[str, dict]) -> None:
    """Persist the remote URL cache."""
    _CACHE_DIR.mkdir(parents=True, exist_ok=True)
    _REMOTE_CACHE_FILE.write_text(json.dumps(cache))


def _git_config_mtime(repo_path: Path) -> float:
    """Return mtime of .git/config (our invalidation key)."""
    config = repo_path / ".git" / "config"
    try:
        return config.stat().st_mtime
    except OSError:
        return 0.0


_CACHE_MISS = object()


def _resolve_remote_cached(
    repo_path: Path, cache: dict[str, dict], now: float
) -> str | None | object:
    """Look up remote URL from cache. Returns _CACHE_MISS on miss."""
    key = str(repo_path.resolve())
    entry = cache.get(key)
    if entry is None:
        return _CACHE_MISS
    if entry.get("mtime", 0) != _git_config_mtime(repo_path):
        return _CACHE_MISS
    if now - entry.get("ts", 0) > _CACHE_TTL_SECONDS:
        return _CACHE_MISS
    url = entry.get("url", "")
    return url or None  # "" → None


def _batch_resolve_remotes(
    repo_paths: list[Path],
) -> dict[Path, str | None]:
    """Resolve remote URLs for many repos, using cache + thread parallelism."""
    cache = _load_remote_cache()
    now = time.time()

    results: dict[Path, str | None] = {}
    to_fetch: list[Path] = []

    for p in repo_paths:
        cached = _resolve_remote_cached(p, cache, now)
        if cached is not _CACHE_MISS:
            results[p] = cached
        else:
            to_fetch.append(p)

    if to_fetch:
        # Parallel git subprocess calls — big speedup for cold cache
        with ThreadPoolExecutor(max_workers=16) as pool:
            futures = {pool.submit(remote_url, p): p for p in to_fetch}
            for future in as_completed(futures):
                p = futures[future]
                try:
                    url = future.result()
                except Exception:
                    url = None
                results[p] = url
                cache[str(p.resolve())] = {
                    "url": url or "",
                    "mtime": _git_config_mtime(p),
                    "ts": now,
                }

        _save_remote_cache(cache)

    return results


# ---------------------------------------------------------------------------
# Public types / helpers
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class RepoInfo:
    """A discovered repository with its identity and location."""

    name: str  # folder name
    path: Path  # absolute path to repo
    remote: str | None  # origin remote URL
    display_name: str  # org/repo from remote, or folder name as fallback


def _display_name_from_url(url: str | None, fallback: str) -> str:
    """Derive a display name from a remote URL, falling back to *fallback*."""
    if url:
        parsed = parse_remote_name(url)
        if parsed:
            return parsed
    return fallback


def _repo_display_name(path: Path) -> str:
    """Derive a display name from the remote URL, falling back to folder name."""
    url = remote_url(path)
    if url:
        parsed = parse_remote_name(url)
        if parsed:
            return parsed
    return path.name


# ---------------------------------------------------------------------------
# Repo discovery
# ---------------------------------------------------------------------------


def find_repos(repos_dir: Path) -> dict[str, Path]:
    """Find all git repos in the given directory (one level deep).

    Returns a dict mapping repo name -> repo path.
    Uses ``.git`` existence check instead of ``git rev-parse`` to avoid
    spawning a subprocess per directory.
    """
    repos: dict[str, Path] = {}
    if not repos_dir.is_dir():
        return repos
    for entry in sorted(repos_dir.iterdir()):
        if entry.is_dir() and not entry.name.startswith(".") and (entry / ".git").is_dir():
            repos[entry.name] = entry
    return repos


def find_all_repos(repo_dirs: list[Path]) -> dict[str, Path]:
    """Find all git repos across multiple directories (one level deep each).

    Returns a dict mapping repo name -> repo path. If the same repo name
    appears in multiple directories, the first occurrence wins.
    """
    repos: dict[str, Path] = {}
    for d in repo_dirs:
        for name, path in find_repos(d).items():
            if name not in repos:
                repos[name] = path
    return repos


def discover_repos(repo_dirs: list[Path], max_depth: int = 3) -> list[RepoInfo]:
    """Deep-scan directories for git repos, deduped by remote URL.

    Returns a list of :class:`RepoInfo` sorted by display name.
    When multiple paths share the same remote URL, the one closest to a
    configured ``repo_dirs`` root wins (i.e. direct children are preferred
    over nested repos).
    """
    # Phase 1: fast filesystem scan (no subprocesses)
    candidates: list[tuple[Path, Path]] = []  # (entry, root)
    seen_paths: set[Path] = set()

    for d in repo_dirs:
        _collect_repo_paths(d, d, candidates, seen_paths, depth=0, max_depth=max_depth)

    if not candidates:
        return []

    # Phase 2: batch-resolve remote URLs (cached + parallel)
    repo_paths = [entry for entry, _ in candidates]
    url_map = _batch_resolve_remotes(repo_paths)

    # Phase 3: build RepoInfo list with dedup
    seen_remotes: dict[str, RepoInfo] = {}
    for entry, root in candidates:
        url = url_map.get(entry)
        display = _display_name_from_url(url, entry.name)
        info = RepoInfo(
            name=entry.name,
            path=entry,
            remote=url,
            display_name=display,
        )

        if url and url in seen_remotes:
            existing = seen_remotes[url]
            if existing.path.parent != root and entry.parent == root:
                seen_remotes[url] = info
        else:
            key = url or str(entry.resolve())
            if key not in seen_remotes:
                seen_remotes[key] = info

    return sorted(seen_remotes.values(), key=lambda r: r.display_name)


def _collect_repo_paths(
    directory: Path,
    root: Path,
    candidates: list[tuple[Path, Path]],
    seen_paths: set[Path],
    depth: int,
    max_depth: int,
) -> None:
    """Recursively collect repo paths without resolving remotes."""
    if depth > max_depth or not directory.is_dir():
        return
    try:
        entries = sorted(directory.iterdir())
    except PermissionError:
        return
    for entry in entries:
        if not entry.is_dir() or entry.name.startswith("."):
            continue
        if (entry / ".git").is_dir():
            resolved = entry.resolve()
            if resolved not in seen_paths:
                seen_paths.add(resolved)
                candidates.append((entry, root))
            # Don't descend into repos
        else:
            _collect_repo_paths(entry, root, candidates, seen_paths, depth + 1, max_depth)


def explore_repos(repo_dirs: list[Path], max_depth: int = 3) -> dict[Path, dict[str, Path]]:
    """Scan directories recursively for git repos, grouped by source dir.

    Returns ``{source_dir: {repo_name: repo_path}}``.
    Scans up to *max_depth* levels deep.  Stops descending into a directory
    once a ``.git`` is found (a repo's subdirs are not scanned).
    """
    result: dict[Path, dict[str, Path]] = {}
    for d in repo_dirs:
        found: dict[str, Path] = {}
        _scan_recursive(d, found, depth=0, max_depth=max_depth)
        if found:
            result[d] = found
    return result


def _scan_recursive(directory: Path, found: dict[str, Path], depth: int, max_depth: int) -> None:
    """Recursively scan *directory* for git repos up to *max_depth*."""
    if depth > max_depth or not directory.is_dir():
        return
    try:
        entries = sorted(directory.iterdir())
    except PermissionError:
        return
    for entry in entries:
        if not entry.is_dir() or entry.name.startswith("."):
            continue
        if (entry / ".git").is_dir():
            name = entry.name
            if name not in found:
                found[name] = entry
            # Don't descend into repos
        else:
            _scan_recursive(entry, found, depth + 1, max_depth)

File: grove-go/internal/discover/discover.go
// Package discover handles repository discovery across configured directories.
package discover

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/nicksenap/grove/internal/config"
	"github.com/nicksenap/grove/internal/git"
	"github.com/nicksenap/grove/internal/models"
)

const (
	remoteCacheTTL = 24 * time.Hour
	maxWorkers     = 16
)

// remoteCacheEntry is a single cached remote URL result.
type remoteCacheEntry struct {
	Remote string `json:"remote"`
	Mtime  int64  `json:"mtime"`
}

// remoteCacheFile is the on-disk format for ~/.grove/cache/remotes.json.
type remoteCacheFile struct {
	Entries   map[string]remoteCacheEntry `json:"entries"`
	UpdatedAt int64                       `json:"updated_at"`
}

func remoteCachePath() string {
	return filepath.Join(config.GroveDir(), "cache", "remotes.json")
}

func loadRemoteCache() *remoteCacheFile {
	data, err := os.ReadFile(remoteCachePath())
	if err != nil {
		return &remoteCacheFile{Entries: make(map[string]remoteCacheEntry)}
	}
	var cache remoteCacheFile
	if err := json.Unmarshal(data, &cache); err != nil {
		return &remoteCacheFile{Entries: make(map[string]remoteCacheEntry)}
	}
	if cache.Entries == nil {
		cache.Entries = make(map[string]remoteCacheEntry)
	}

	// Check TTL
	if time.Since(time.Unix(cache.UpdatedAt, 0)) > remoteCacheTTL {
		return &remoteCacheFile{Entries: make(map[string]remoteCacheEntry)}
	}
	return &cache
}

func saveRemoteCache(cache *remoteCacheFile) {
	cache.UpdatedAt = time.Now().Unix()
	data, err := json.MarshalIndent(cache, "", "  ")
	if err != nil {
		return
	}
	dir := filepath.Dir(remoteCachePath())
	os.MkdirAll(dir, 0o755)
	os.WriteFile(remoteCachePath(), data, 0o644)
}

// FindRepos discovers repos one level deep in a directory.
func FindRepos(repoDir string) map[string]string {
	repos := make(map[string]string)
	entries, err := os.ReadDir(repoDir)
	if err != nil {
		return repos
	}
	for _, entry := range entries {
		if !entry.IsDir() || strings.HasPrefix(entry.Name(), ".") {
			continue
		}
		full := filepath.Join(repoDir, entry.Name())
		if git.IsGitRepo(full) {
			repos[entry.Name()] = full
		}
	}
	return repos
}

// FindAllRepos discovers repos across all configured directories, deduplicating by name.
func FindAllRepos(repoDirs []string) map[string]string {
	repos := make(map[string]string)
	for _, dir := range repoDirs {
		for name, path := range FindRepos(dir) {
			if _, exists := repos[name]; !exists {
				repos[name] = path
			}
		}
	}
	return repos
}

// DiscoverRepos performs a deep scan with deduplication by remote URL.
func DiscoverRepos(repoDirs []string, maxDepth int) []models.RepoInfo {
	if maxDepth <= 0 {
		maxDepth = 3
	}

	// Phase 1: Fast filesystem scan
	var candidates []candidate
	for _, dir := range repoDirs {
		collectRepoPaths(dir, 0, maxDepth, &candidates)
	}

	// Phase 2: Resolve remotes (cached + parallel)
	cache := loadRemoteCache()
	type resolvedRepo struct {
		name    string
		path    string
		remote  string
		display string
		depth   int
	}

	var (
		resolved []resolvedRepo
		mu       sync.Mutex
		wg       sync.WaitGroup
		sem      = make(chan struct{}, maxWorkers)
	)

	for _, c := range candidates {
		wg.Add(1)
		go func(c candidate) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()

			remote := ""
			display := c.name

			// Check cache by .git/config mtime
			gitConfig := filepath.Join(c.path, ".git", "config")
			if info, err := os.Stat(gitConfig); err == nil {
				mtime := info.ModTime().Unix()
				if cached, ok := cache.Entries[c.path]; ok && cached.Mtime == mtime {
					remote = cached.Remote
				} else {
					if url, err := git.RemoteURL(c.path, "origin"); err == nil {
						remote = url
					}
					mu.Lock()
					cache.Entries[c.path] = remoteCacheEntry{Remote: remote, Mtime: mtime}
					mu.Unlock()
				}
			}

			if remote != "" {
				if parsed := git.ParseRemoteName(remote); parsed != "" {
					display = parsed
				}
			}

			// Calculate depth from any repo_dir
			depth := 999
			for _, dir := range repoDirs {
				rel, err := filepath.Rel(dir, c.path)
				if err == nil {
					d := strings.Count(rel, string(filepath.Separator))
					if d < depth {
						depth = d
					}
				}
			}

			mu.Lock()
			resolved = append(resolved, resolvedRepo{
				name:    c.name,
				path:    c.path,
				remote:  remote,
				display: display,
				depth:   depth,
			})
			mu.Unlock()
		}(c)
	}
	wg.Wait()
	saveRemoteCache(cache)

	// Phase 3: Dedup by remote, prefer shallower paths
	sort.Slice(resolved, func(i, j int) bool {
		return resolved[i].depth < resolved[j].depth
	})

	seen := make(map[string]bool)
	var result []models.RepoInfo
	for _, r := range resolved {
		key := r.remote
		if key == "" {
			key = r.path // No remote → use path as unique key
		}
		if seen[key] {
			continue
		}
		seen[key] = true
		result = append(result, models.RepoInfo{
			Name:        r.name,
			Path:        r.path,
			Remote:      r.remote,
			DisplayName: r.display,
		})
	}

	sort.Slice(result, func(i, j int) bool {
		return result[i].DisplayName < result[j].DisplayName
	})
	return result
}

// ExploreRepos groups discovered repos by source directory.
func ExploreRepos(repoDirs []string, maxDepth int) map[string]map[string]string {
	if maxDepth <= 0 {
		maxDepth = 3
	}
	grouped := make(map[string]map[string]string)
	for _, dir := range repoDirs {
		repos := make(map[string]string)
		walkRepos(dir, 0, maxDepth, repos)
		if len(repos) > 0 {
			grouped[dir] = repos
		}
	}
	return grouped
}

func walkRepos(dir string, depth, maxDepth int, repos map[string]string) {
	if depth >= maxDepth {
		return
	}
	entries, err := os.ReadDir(dir)
	if err != nil {
		return
	}
	for _, entry := range entries {
		if !entry.IsDir() || strings.HasPrefix(entry.Name(), ".") {
			continue
		}
		full := filepath.Join(dir, entry.Name())
		if git.IsGitRepo(full) {
			repos[entry.Name()] = full
		} else {
			walkRepos(full, depth+1, maxDepth, repos)
		}
	}
}

type candidate = struct {
	name string
	path string
}

func collectRepoPaths(dir string, depth, maxDepth int, out *[]candidate) {
	if depth >= maxDepth {
		return
	}
	entries, err := os.ReadDir(dir)
	if err != nil {
		return
	}
	for _, entry := range entries {
		if !entry.IsDir() || strings.HasPrefix(entry.Name(), ".") {
			continue
		}
		full := filepath.Join(dir, entry.Name())
		if git.IsGitRepo(full) {
			if out != nil {
				*out = append(*out, candidate{name: entry.Name(), path: full})
			}
		} else {
			collectRepoPaths(full, depth+1, maxDepth, out)
		}
	}
}
